.github/scripts/main/main.sh

```diff
@@ -121,35 +163,88 @@ if [ "$OPAM_TEST" = "1" ]; then
-    OPAM_RT_BRANCH=$GITHUB_BASE_REF
-  else
-    OPAM_RT_BRANCH=master
-  fi
-  if git branch | grep -q "$OPAM_RT_BRANCH"; then
-    git checkout "$OPAM_RT_BRANCH"
-    git reset --hard "origin/$OPAM_RT_BRANCH"
-  else
-    git checkout -b "$OPAM_RT_BRANCH" "origin/$OPAM_RT_BRANCH"
-  fi
+  prepare_project "https://github.com/ocaml-opam/opam-rt" "opam-rt"
 
-  test -d _opam || opam switch create . --no-install --formula '"ocaml-system"'
-  eval $(opam env)
-  opam pin $GITHUB_WORKSPACE -yn --with-version to-test
   # opam lib pins defined in opam-rt are ignored as there is a local pin
   opam pin . -yn --ignore-pin-depends
-  opam install opam-rt --deps-only opam-devel.to-test
-  make || { opam reinstall opam-client -y; make; }
+  opam install opam-rt --deps-only opam-devel
+  opam exec -- make || { opam reinstall opam-client -y; opam exec -- make; }
   (set +x ; echo -en "::endgroup::opam-rt\r") 2>/dev/null
 fi
+
+test_project () {
+  url=$1
+  project=$2
+
+  (set +x; echo -en "::group::depends-$project\r") 2>/dev/null
+  set +e
+  opam pin "$url" --kind git -yn
+  for pkg_name in $(opam show . -f name); do
+    echo "Installing dependencies for $pkg_name"
+    opam install "$pkg_name" --deps-only
+    deps_code=$?
+    if [ $deps_code -ne 0 ]; then
+      echo "Dependency installation failed for $pkg_name"
+      DEPENDS_ERRORS="$DEPENDS_ERRORS $pkg_name"
+    else
+      echo "Installing opam-client and $pkg_name"
+      opam install opam-client
+      opam install "$pkg_name"
+      code=$?
+      if [ $code -ne 0 ]; then
+        LIB_ERRORS="$LIB_ERRORS $project"
+        echo -e "\e[31mErrors while installing $pkg_name\e[0m";
+      fi
+    fi
+  done
+  set -e
+  (set +x ; echo -en "::endgroup::depends-$project\r") 2>/dev/null
+}
+
+if [ "$OPAM_DEPENDS" = "1" ]; then
+  DEPENDS_ERRORS=""
+  LIB_ERRORS=""
+  OCAMLVER=$(ocamlc -version)
+
+  (set +x; echo -en "::group::depends\r") 2>/dev/null
+  eval $(grep '^VERSION=' shell/install.sh)
+  opam_libs=$(opam show . -f name 2>/dev/null)
+  depends_on=$(echo "$opam_libs" | sed "s/\$/.${VERSION}/" | paste -sd, -)
+  packages=$(opam list --or --depends-on "$depends_on" --columns name | tail -n +3)
+  set +x
+  for exclude in $opam_libs; do
+    packages=$(echo "$packages" | grep -v -x "$exclude")
+  done
+  set -x
+
+  for pkg in $packages; do
+    dev_repo=$(opam show "$pkg" -f dev-repo 2>/dev/null)
+    dev_repo=$(echo "$dev_repo" | sed -E 's/^"//;s/"$//;s/^git\+//;s/\.git$//')
+
+    if [[ -n "$dev_repo" ]]; then
+      prepare_project "$dev_repo" "$pkg"
+      test_project "$dev_repo" "$pkg"
+    fi
+  done
+
+  if [ -n "$DEPENDS_ERRORS" ]; then
+    echo -e "\e[31mErrors detected in dependencies of $DEPENDS_ERRORS\e[0m";
+  fi
+
+  if [ -n "$LIB_ERRORS" ]; then
+    FAIL=""
+    set +x
+    for critical in $FAIL_IF_DEPENDENT; do
+      if echo "$LIB_ERRORS" | grep -wq "$critical"; then
+        FAIL="$FAIL $critical"
```

Comment:
It doesn't really matter here, but note that bash (not posix sh) has generic arrays that are a little easier to use for this kind of thing
```suggestion
        FAIL+=("$critical")
```